File: Scala/src/main/scala/Ackermann.scala
object Ackermann {

  def Ack(m:BigInt, n:BigInt): BigInt ={

    if (m == 0) n + 1

    else if (n == 0) Ack(m - 1, 1)

    else Ack(m - 1, Ack(m, n - 1))

  }

  def main(args: Array[String]):Unit = {

    val inicio = System.nanoTime()

    // val resultado = Ack(1, 1)
    // val resultado = Ack(1, 2)
     val resultado = Ack(2, 1)
    // val resultado = Ack(2, 2)
    // val resultado = Ack(3, 1)
    // val resultado = Ack(3, 2)
    // val resultado = Ack(3, 2)

    val fim = System.nanoTime()

    println(resultado)

    println(fim - inicio)

    println(Runtime.getRuntime.totalMemory - Runtime.getRuntime.freeMemory)

  }

}

File: Java/src/com/senai/Ackermann.java
package com.senai;

import java.math.BigInteger;

class Ackermann {

    public static BigInteger Ack(BigInteger m, BigInteger n){

        if(m.equals(BigInteger.ZERO)) {

            return n.add(BigInteger.ONE);

        } else if(n.equals(BigInteger.ZERO)) {

            return Ack(m.subtract(BigInteger.ONE), BigInteger.valueOf(1));

        } else {

            return Ack(m.subtract(BigInteger.ONE), Ack(m, n.subtract(BigInteger.ONE)));

        }

    }

    public static void main(String args[]) {

        long inicio = System.nanoTime();

        // BigInteger resultado = Ack(BigInteger.valueOf(1), BigInteger.valueOf(1));
        // BigInteger resultado = Ack(BigInteger.valueOf(1), BigInteger.valueOf(2));
         BigInteger resultado = Ack(BigInteger.valueOf(2), BigInteger.valueOf(1));
        // BigInteger resultado = Ack(BigInteger.valueOf(2), BigInteger.valueOf(2));
        // BigInteger resultado = Ack(BigInteger.valueOf(3), BigInteger.valueOf(1));
        // BigInteger resultado = Ack(BigInteger.valueOf(3), BigInteger.valueOf(2));
        //BigInteger resultado = Ack(BigInteger.valueOf(3), BigInteger.valueOf(2));

        long fim = System.nanoTime();

        System.out.println(resultado);

        System.out.println(fim - inicio);

        System.out.println(Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory());

    }

}